PR to fix a number of small issues

valeriupredoi opened the pull request:
This should fix the following issues so far:
 - https://github.com/ESMValGroup/ESMValTool/issues/851
 - https://github.com/ESMValGroup/ESMValTool/issues/819
 - does not solve #810 but I put the memory equations in the documentation (btw #810 can not be solved immediately, it will be a long and rocky road that will go through a town called Dask)

NOTE to devs: please commit to this PR if the issue is __SMALL__ and can be solved by no more than 3-4 +/-

bouweandela commented:
Issue #819 is mostly a missing documentation issue, so in order to close it, some documentation will need to be written.

bouweandela commented on esmvaltool/preprocessor/_regrid.py:
```diff
@@ -508,8 +506,8 @@ def get_reference_levels(filename,
     cube = cubes[0]
     try:
-        coord = cube.coord(coordinate)
+        coord = cube.coord(coord)
```
```suggestion
        coord = cube.coord(axis='Z')
```
and remove the function argument

valeriupredoi replied:
yes, you are falling for the same I did yesterday and reverted the changes, the tests are done with a non-descriptive z axis, and if you change the test a whole lot of other crap starts falling; just leave it this way and the user can give it any (working) value

bouweandela replied:
Just fix the tests instead, cmorized data should have the correct z axis, right?

valeriupredoi replied:
I think so but am not 100% sure - @jvegasbsc ?

bouweandela commented:
I think the time slicing checks may be a bit too strict and lead to unnecessary errors. How about just raising the exception if no data at all was available in the file?

valeriupredoi commented:
> I think the time slicing checks may be a bit too strict and lead to unnecessary errors. How about just raising the exception if no data at all was available in the file?

what sort of cases do you have in mind? This check is pretty loose as it is anyway

valeriupredoi commented:
the thing I'm trying to check for is if there is actual data in the file but it's shifted by a lot wrt time slice as it was the case @mattiarighi experienced a few days ago

mattiarighi commented:
> the thing I'm trying to check for is if there is actual data in the file but it's shifted by a lot wrt time slice as it was the case @mattiarighi experienced a few days ago

That was a pretty unlikely case, so if this is too much work you can just forget about it.

bouweandela commented:
> what sort of cases do you have in mind? This check is pretty loose as it is anyway

I've had another look and I think you're right, but the logic is horribly complicated. Why not use a simple check instead, I think it checks the same thing.

valeriupredoi commented:
> but the logic is horribly complicated

I love you too @bouweandela :heart: 

what would you propose :grin:

bouweandela commented:
Just check if any data is left after extracting, if there's nothing left you know that there was nothing within the specified time range: https://github.com/ESMValGroup/ESMValTool/issues/851#issuecomment-462832732

valeriupredoi commented:
@bouweandela proves out that the only way to assign an axis to an (unCMORized) cube coord is to bloody guess it __and__ assign the positive attribute to up - setattr or or simply assigning it when constructing the coord won't do (I spent 1h trying to figure this out!) - just when I was saying iris is quite friendly:

```
+        self.cube.coord('coord').attributes['positive'] = 'up'
+        iris.util.guess_coord_axis(self.cube.coord('coord'))
```

have a look at the changes pls; this is defo more general and open than before

bouweandela commented on esmvaltool/preprocessor/_time_area.py:
```diff
@@ -61,6 +61,8 @@ def time_slice(cube, start_year, start_month, start_day, end_year, end_month,
 
     cube_slice = cube.extract(constraint)
+    if cube_slice is None:
+        raise ValueError("Time slice outside cube time boundaries.")
```
Maybe it would be nice to mention the values of start and stop time requested and what is available in the cube?

valeriupredoi replied:
will do now

bouweandela commented:
> have a look at the changes pls; this is defo more general and open than before

This looks more user friendly. Maybe @ledm can test if it solves his issue #819?

valeriupredoi commented:
partially addresses #869 pending OK from @ledm and if the ok comes then I'll write a unit test for it and will fully address that issue; and that's it - no more changes here - :checkered_flag:

valeriupredoi commented:
@ledm you have a nice lovely preprocessor function called `_align_time_axes()` that works well and has a strong unit test now - to solve #869 :grin: I'd be very grateful if you could test and see if it suites the needs; also note that I didn't yet give it a user-friendly name in preprocessor/init coz I left that for you to call it whatevs, also the unit test imports the raw func name so when you baptize it change that too pls

valeriupredoi commented:
ok hustle up peeps, it'd be nice if we tested&approved&closed this to close the associated issues :grin:

ledm commented:
The more changes we make to `_time_area.py`, `_area_pp.py` and `_volume_pp.py`, the harder it will be to merge #875 & #876. (That PR renames these files.)

Issue #847 & PR #875 are tested and ready to go right now. If they get merged in first, then this PR will need to be re-worked and re--tested. This is why I hesitate to run any tests for this PR now as I'll have to repeat them again later.

valeriupredoi commented:
OK let's wait for #875

ledm commented:
Hmmm... this one is still waiting for #875...

valeriupredoi commented:
#875 has been merged; can ppl have a look at this one pls? @ledm @bouweandela @mattiarighi @jvegasbsc :beer:

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -61,6 +62,14 @@ def extract_time(cube, start_year, start_month, start_day, end_year, end_month,
+    if cube_slice is None:
+        start_cube = str(cube.coord('time').points[0])
+        end_cube = str(cube.coord('time').points[-1])
+        raise ValueError("Time slice {} to {} is outside \
```
Don't use backslashes, strings are automatically concatenated when they are between brackets, e.g. `raise ValueError("a" "b")` will work just fine.

bouweandela replied:
In fact, now that we're dropping Python 2, you could even use an [f-string](https://docs.python.org/3/whatsnew/3.6.html#pep-498-formatted-string-literals):
```
raise ValueError(
    f"Time slice {start_date} to {end_date} is outside cube time bounds "
    f"{start_cube} to {end_cube}.")
```

valeriupredoi replied:
getting a syntax error with the f-string (lovely name btw :grin: ) but guessing my propector is a bit old

valeriupredoi replied:
actually CodaCI doesn't like it either

valeriupredoi replied:
ow wait that coz py2 integration is still on for this pr, nevermind

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -162,10 +171,68 @@ def seasonal_mean(cube):
+                          calendar=cube.coord('time').units.calendar))
+
+        # fix calendars
+        cube.coord('time').units = \
```
Don't use backslashes

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -162,10 +171,68 @@ def seasonal_mean(cube):
+        # standardize time points
+        time_c = [cell.point for cell in cube.coord('time').cells()]
+        cube.coord('time').cells = [
+            datetime.datetime(t.year, t.month, 15, 0, 0, 0) for t in time_c
```
It looks like this only works for monthly data, what about daily, hourly etc?

bouweandela replied:
Also, wouldn't a choice of regridding schemes, like for the horizontal and vertical regridders, make more sense?

valeriupredoi replied:
mmyeahh...I added a few more cases kwarged on `frequency`; not sure about other types of data since I haven't yet worked with them; no man, no need for time-regridding, that would be overly complicated, all the cubes need are common points, same aux coords and the same bounds

bouweandela replied:
But that would be inaccurate, right? @mattiarighi @ledm Do you know if it would make sense to support proper temporal regridding? I think this could be a useful feature for preprocessing data going into computing multimodel statistics too.

mattiarighi replied:
As we discussed in another PR, I am a bit skeptical towards temporal regridding.
For monthly mean data that's easy, since the number of timesteps is always the same (12 per year).

But for higher time frequency it is calendar-dependent and we need to make assumptions on how to set a common time coordinate for all datasets when calendars are differing.

I also think that a non-standard calendar (like 360) is not suitable for temporally colocating model and observations on a daily or hourly basis, but just monthly.

bouweandela replied:
> For monthly mean data that's easy, since the number of timesteps is always the same (12 per year).

Yes, but I could image that it makes a big difference if the time point is the first of the month vs the middle or even last day of the month, right? Wouldn't it make sense to use linear interpolation for that case, if you want to compare data from the 1st of the month to the 15th?

valeriupredoi replied:
currently the function sets it to the 15th of the month as common point. Why would it make a difference if it is the first or the middle of the month? Issues appear if it's 29th of the month onwards because February

bouweandela commented on tests/unit/preprocessor/_time/test_time.py:
```diff
@@ -13,7 +13,8 @@
 from esmvaltool.preprocessor._time import (extract_month, extract_season,
-                                           time_average, extract_time)
+                                           time_average, extract_time,
+                                           _align_time_axes)
```
use isort

bouweandela replied:
This is not done yet

valeriupredoi replied:
it is, the web editor is wonk, the code is:
```
from esmvaltool.preprocessor._time import (_align_time_axes, extract_month,
                                                                      extract_season, extract_time,
                                                                      time_average)
```

on a related topic it is worth telling people to keep isort updated since it is a tool outside the esmvaltool environment

valeriupredoi replied:
behh, don't mind the formatting on my code snippet, I am stoopid:
```
from esmvaltool.preprocessor._time import (_align_time_axes, extract_month,
                                           extract_season, extract_time,
                                           time_average)
```

valeriupredoi replied:
aaaand only now it's fixed :grin:

bouweandela replied:
> on a related topic it is worth telling people to keep isort updated since it is a tool outside the esmvaltool environment

Not if you install esmvaltool with `pip install -e .[develop]` as recommended in CONTRIBUTING.md

valeriupredoi commented:
I agree with Mattia. This is a time aligner not a regridder, this assumes
the number of time points is the same for all the cubes involved. Time
regridding on cubes with different calendars means we need to extrapolate
data for missing days (like no leap or 360 day calendars) and that could
lead to large biases if the analysis is done on small time intervals

Dr Valeriu Predoi.
Computational scientist
NCAS-CMS
University of Reading
Department of Meteorology
Reading RG6 6BB
United Kingdom

On Fri, 8 Mar 2019, 09:32 Mattia Righi, <notifications@github.com> wrote:

> *@mattiarighi* commented on this pull request.
> ------------------------------
>
> In esmvaltool/preprocessor/_time.py
> <https://github.com/ESMValGroup/ESMValTool/pull/867#discussion_r263715785>
> :
>
> > +        # fix units; leave calendars
> +        # this is not normally needed unless working
> +        # outside the CMOR standards
> +        cube.coord('time').convert_units(
> +            cf_units.Unit('days since 1950-1-1 00:00:00',
> +                          calendar=cube.coord('time').units.calendar))
> +
> +        # fix calendars
> +        cube.coord('time').units = \
> +            cf_units.Unit(cube.coord('time').units.origin,
> +                          calendar='gregorian')
> +
> +        # standardize time points
> +        time_c = [cell.point for cell in cube.coord('time').cells()]
> +        cube.coord('time').cells = [
> +            datetime.datetime(t.year, t.month, 15, 0, 0, 0) for t in time_c
>
> As we discussed in another PR, I am a bit skeptical towards temporal
> regridding.
> For monthly mean data that's easy, since the number of timesteps is always
> the same (12 per year).
>
> But for higher time frequency it is calendar-dependent and we need to make
> assumptions on how to set a common time coordinate for all datasets when
> calendars are differing.
>
> I also think that a non-standard calendar (like 360) is not suitable for
> temporally colocating model and observations on a daily or hourly basis,
> but just monthly.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/867#discussion_r263715785>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCo202rt9gxOgVe0I-Q7nnEnZZXnzWks5vUi4qgaJpZM4bBJ3S>
> .
>

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -162,10 +169,84 @@ def seasonal_mean(cube):
+    for cube in cubes:
+        # fix units; leave calendars
+        # this is not normally needed unless working
+        # outside the CMOR standards
```
Preprocessor functions are intended for use with CMORized data, so remove

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -162,10 +169,84 @@ def seasonal_mean(cube):
                               iris.analysis.MEAN)
 
-    # TODO: This preprocessor is not calendar independent.
+    # this func returns an approximation to 3 months
```
This function has an open issue (cannot find it at the moment), leave the TODO

bouweandela replied:
Found it: https://github.com/ESMValGroup/ESMValTool/issues/664

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -162,10 +169,84 @@ def seasonal_mean(cube):
+                datetime.datetime(t.year, t.month,
+                                  15, 0, 0, 0) for t in time_c
+            ]
+        if frequency == 'daily':
```
```suggestion
        elif frequency == 'daily':
```

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -162,10 +169,84 @@ def seasonal_mean(cube):
+
+        # standardize time points
+        time_c = [cell.point for cell in cube.coord('time').cells()]
+        if frequency == 'monthly':
```
Add unit tests for all cases

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -162,10 +169,84 @@ def seasonal_mean(cube):
+                datetime.datetime(t.year, t.month,
+                                  t.day, 0, 0, 0) for t in time_c
+            ]
+        if frequency == 'hourly':
```
```suggestion
        elif frequency == 'hourly':
```

valeriupredoi commented:
@bouweandela man, addressed your comments, cheers, brother! Note that I actually removed the `hourly` case for time alignment - that one is quite tricky and needs more attention, `monthly` and `daily` have their own tests and should work fine :ok_hand:

valeriupredoi commented:
@ledm @bouweandela are we happies with this?

bouweandela commented on tests/unit/preprocessor/_time/test_time.py:
```diff
@@ -36,7 +49,6 @@ def setUp(self):
     def test_get_january(self):
         """Test january extraction"""
         sliced = extract_month(self.cube, 1)
-        print(sliced)
```
Put the `print` functions back, they help debugging if the test fails.

valeriupredoi replied:
ok since the print call will print only if it fails

valeriupredoi replied:
done

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -183,14 +190,80 @@ def seasonal_mean(cube):
     return cube.extract(three_months_bound)
 
 
+def _align_time_axes(cubes, frequency='monthly'):
```
Would it be possible to figure out `frequency` from the cubes, instead of making the user of the function provide it? This function also needs a better name. Also, it should be documented that this function only works for monthly and daily data.

valeriupredoi replied:
yes, from CMIP6 it is readily available from variable, but it's a bit of a hassle to get it out cleanly, I'll see what I can sort out

valeriupredoi replied:
OK I gave it a thought and it's not straightforward since the only piece of info here is coded in `cube` that may or may not have any info on its frequency so I think we'll leave it as func arg

valeriupredoi replied:
how would you rename it? (I'll rename it first then add the dox info)

bouweandela replied:
It should be pretty straightforward to figure out the frequency from the time data points. If you have only two or more points, just check if the difference is more than 1, in that case it's monthly data, else daily. If you have only a single point, I guess you can just assign the same time point to every cube.

bouweandela replied:
Maybe call it `regrid_time`?

valeriupredoi replied:
yeah sounds good, only that it doesnt actually regrid in the interpolative sense :grin:

valeriupredoi replied:
> It should be pretty straightforward to figure out the frequency from the time data points. If you have two or more points, just check if the difference equals 1, in that case it's daily data, else monthly. If you have only a single point, I guess you can just assign the same time point to every cube.

ooh no man, that is kludgy, I'd stay away from in-place verification specially if we want to extend this to hourly and extend the CMOR units from days only and support other time units; I'd leave it like this for now, this function is easily usable both from the recipe and diag and has only one argument anyway :grin:

bouweandela replied:
In that case you can add something to _recipe.py so it is at least filled in automatically? Similar to how the info for other regridding functions is suppplied?

valeriupredoi replied:
sounds like a good plan, will do this now

valeriupredoi replied:
hey @bouweandela so I just made `regrid_time` a valid preproc function that can take `frequency` as argument (if user desires)

bouweandela replied:
Yes, I saw, very good. But if you want to use it, you have to specify `frequency` in the recipe. For other preprocessor functions, that information is filled in automatically by the recipe parser.

valeriupredoi replied:
you can run it without `frequency` and that will pick up the default `monthly`; I can the automated frequency inputter based on mip but I think this should be left to the user to decide - what you think?

bouweandela replied:
frequency is read from the cmor table, it is already available in the variable dicts in recipe.py, you should just add it to the settings of this function (if it's enabled in the recipe).

valeriupredoi replied:
you are absolutely correct and I should have thought of this since I actually bloody made it work in `table.py` for the newest CMIP6 tables - good one, man :beer:

valeriupredoi replied:
le done monsieur le Bouwe :grin:

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -61,6 +62,12 @@ def extract_time(cube, start_year, start_month, start_day, end_year, end_month,
             time=lambda t: (t_1 < time_units.date2num(t.point) < t_2))
 
     cube_slice = cube.extract(constraint)
+    if cube_slice is None:
```
Add a unit test for this

valeriupredoi replied:
done

valeriupredoi commented:
@bouweandela man all a-ok now?

mattiarighi commented:
Will test it once approved.

bouweandela commented on esmvaltool/_recipe.py:
```diff
@@ -210,6 +208,16 @@ def _update_target_grid(variable, variables, settings, config_user):
 
+def _update_regrid_time(variable, settings):
+    """Input data frequency automatically for regrid_time preprocessor."""
+    regrid_time = settings.get('regrid_time')
```
```suggestion
```

bouweandela commented on esmvaltool/_recipe.py:
```diff
@@ -210,6 +208,16 @@ def _update_target_grid(variable, variables, settings, config_user):
+def _update_regrid_time(variable, settings):
+    """Input data frequency automatically for regrid_time preprocessor."""
+    regrid_time = settings.get('regrid_time')
+    if regrid_time is None:
```
```suggestion
```

bouweandela commented on esmvaltool/_recipe.py:
```diff
@@ -210,6 +208,16 @@ def _update_target_grid(variable, variables, settings, config_user):
+    """Input data frequency automatically for regrid_time preprocessor."""
+    regrid_time = settings.get('regrid_time')
+    if regrid_time is None:
+        return
```
```suggestion
```
this is not needed, I think

valeriupredoi replied:
it is, because I want to keep the user's choice of passing in `frequency` in recipe, without it if there is no `regrid_time` in recipe, the code returns a `KeyError`

bouweandela replied:
Try it first

valeriupredoi replied:
I did! It's also a bitty bit obvious no? :grin:

bouweandela replied:
I was sleeping ;-)

valeriupredoi replied:
that's okay, B-man, catch some sleep you missed from home :grin:

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -192,6 +199,73 @@ def spans_three_months(time):
+    coordinates so that any cube from cubes can be subtracted from any
+    other cube from cubes. Currently this function supports only monthly
+    and daily data time frequencies, but it will be adapted for sub-day
+    frequencies in the future. Frequency is passed in straight from
```
Remove the information 'Frequency is passed in straight from ..`, this is relevant only if using the function from the recipe, not when you use it as a python function.

bouweandela replied:
Also, remove the ', but it will be adapted', we can see about that when you actually do it.

bouweandela commented on esmvaltool/preprocessor/_time.py:
```diff
@@ -192,6 +199,73 @@ def spans_three_months(time):
+        cube: iris.cube.Cube
+        frequency: str
+            data frequency: mon or day
+            default: monthly
```
```suggestion
```

valeriupredoi commented:
arite B-man, let's wrap this up shall we :beer:  @bouweandela

mattiarighi commented:
Will test in a few minutes...
